zerver/actions/message_flags.py

```diff
@@ -44,15 +44,26 @@ def do_mark_all_as_read(user_profile: UserProfile) -> int:
     )
     do_clear_mobile_push_notifications_for_ids([user_profile.id], all_push_message_ids)
 
-    with transaction.atomic(savepoint=False):
-        query = (
-            UserMessage.select_for_update_query()
-            .filter(user_profile=user_profile)
-            .extra(where=[UserMessage.where_unread()])
-        )
-        count = query.update(
-            flags=F("flags").bitor(UserMessage.flags.read),
-        )
+    batch_size = 10000
+    count = 0
+    while True:
+        with transaction.atomic(savepoint=False):
+            query = (
+                UserMessage.select_for_update_query()
+                .filter(user_profile=user_profile)
+                .extra(where=[UserMessage.where_unread()])[:batch_size]
+            )
+            # This updated_count is the same as the number of UserMessage
+            # rows selected, because due to the FOR UPDATE lock, we're guaranteed
+            # that all the selected rows will indeed be updated.
+            # UPDATE queries don't support LIMIT, so we have to use a subquery
+            # to do batching.
+            updated_count = UserMessage.objects.filter(id__in=query).update(
+                flags=F("flags").bitor(UserMessage.flags.read),
+            )
+            count += updated_count
+            if updated_count < batch_size:
+                break
```

Comment:
I think the `do_increment_logging_stat` calls should be inside this transaction block, unlike the `send_event` call.